app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -14,6 +14,7 @@ model: ProvisioningTemplate
 <%= "\n# Location: [#{@location.name}]" if @location -%>
 <%= "\n# Hostgroup: [#{@hostgroup.name}]" if @hostgroup -%>
 <%= "\n# Operating System: [#{@operatingsystem.name}]" if @operatingsystem -%>
+<%= "\n# Setup Remote Execution: [#{@setup_remote_execution}]" -%>
```

Comment:
Does it make sense to implement this as generic feature flags?

So in the template you'd have something like:

```ruby
<% if feature_active?('remote_execution_setup') %>
```

Reply:
Could you please elaborate this idea a bit more? Is that a new object representing some functionality? Would we then automatically render the template for all features? (less readable for users I guess, however much more consistent) I think it would be good for keeping consistency between provisioning and registration templates.

It would be good to add it as a new RFE linked to https://projects.theforeman.org/issues/30440 or start the RFC if you feel it's bigger. I'd like to get this functionality to 2.3, so I hope you don't mind if I merge this without this bigger change.

Reply:
@ares:

> Is that a new object representing some functionality?

Basically, instead of having `setup_insights`, `setup_remote_execution`, `setup_puppet`, `setup_ansible`, `setup_icinga`, `setup_cowsay` and more the idea was to make this a generic interface. A user can specify a comma-separated list via a URL parameter. The template would split this list`params['features'].split(',')` and make it consumable via the macro specified above. This could be a boolean option or even a key/value pair that transfers the key/value pairs to host parameters.
My idea there was, that during traditional provisioning we use host parameters to customize the templates. Every foreman user can specify arbitrary strings here and therefore customize the templates without any rails programming skills. I was thinking about how we can make this possible for registration templates as well.

> I'd like to get this functionality to 2.3, so I hope you don't mind if I merge this without this bigger change.

Well, I do mind merging a PR when we're still in the middle of the conversation. Honestly, this doesn't encourage collaborating on a project. If this change goes into 2.3 changing the behavior later would mean deprecating APIs and is a lot more hassle than iterating on the PR. There are enough open PRs I'd like to have in the next release. But that doesn't stop us from waiting patiently until we reach a consensus.
However, this is probably something we should discuss "in person" and not via Github comments. ;-)

Reply:
Before I open a new Redmine issue for this, let's decide if this is idea actually a good approach or not. What do you think? How far do we want to go in the first iteration?

Reply:
How would the user specify it? I think separate parameters `setup_X` makes it easy to set it on various levels (hostgroup, host) and if it is a single parameter, merging may be harder.

Reply:
This is currently a URL parameter on the global registration endpoint. Correct me if I'm wrong, but this is therefore part of the first `curl` command. I believe specifying `setup_X=true&setup_Y=true` or `features=X,Y` does not make a huge usability difference. The advantage of the latter is, that the possible features aren't hardcoded.

Reply:
First of all, sorry for my misunderstanding. I didn't realize you talk about URL/HTTP parameters in your comment, I thought you talk about variables being organized in features. That would be just internal methods organization, no public API change.

I do like the proposal to organize the setup parameters somehow. Few thoughts
* there's no 1:1 mapping between registration URL parameter to Host (or global or whatever) parameter, so e.g. `setup_ntp_server` shouldn't just update the param, plugins may want to add additional logic
* the list should be dynamic, but it should be easy to add documentation for every added parameter, otherwise user wouldn't know, what all can be customized
* I wonder if we should wrap params under setup key, like we wrap host params, e.g. `setup[x]=...&setup[y]=...`, that would kind of namespace that. OTOH it may be unnecessary complication, just `setup_` prefix could be enough.

This change would be good to get to 2.3 branch, but if we won't have it ready on time, I think it's worth the hassle of deprecating and providing the new way. Given this is a new functionality, it won't be used much yet.

> This is currently a URL parameter on the global registration endpoint. Correct me if I'm wrong, but this is therefore part of the first curl command. I believe specifying setup_X=true&setup_Y=true or features=X,Y does not make a huge usability difference. The advantage of the latter is, that the possible features aren't hardcoded.

I think the parameter needs to be on both endpoints. People need to be able to pass it to GRT (or specify it in UI) so it's then passed to HRT. Only HRT can create host parameters, because only then the host object exists.

Reply:
@ares:
> plugins may want to add additional logic

Can/shouldn't this logic be part of the template (logic)?

> it should be easy to add documentation for every added parameter

In the current provisioning templates, we just do this via a comment `This template accepts the following parameters:`. Is that enough?

> Only HRT can create host parameters

I guess it depends on whether we want this data to be set on the host as parameters or not. What makes this a little more complex is, that these parameters could have been set on the hostgroup already. Just thinking out loud: We could split this PR up and just display the parameters from the hostgroup in the UI and then make them overridable later.